Use Iceberg writers for Parquet data written from Spark.

rdblue opened the pull request:
This removes the use of Spark's Parquet WriteSupport and replaces it with Parquet writers like those used for Iceberg generics and Avro records.

rdblue commented:
@aokolnychyi, this PR avoids using Spark's WriteSupport for Parquet. It would be interesting to see which option performs better.

aokolnychyi commented:
@rdblue I'll give it a try in a few days

aokolnychyi commented:
I have a local benchmark that compares writing 5000000 records in Parquet format using gzip compression via Iceberg and via Spark native file source.

Here are the results:
```
Benchmark                                           Mode  Cnt  Score   Error  Units
SparkParquetFlatDataWriteBenchmark.writeFileSource    ss    5  4.587 ± 0.226   s/op
SparkParquetFlatDataWriteBenchmark.writeIceberg       ss    5  4.537 ± 0.090   s/op
```

```
Benchmark                                             Mode  Cnt  Score   Error  Units
SparkParquetNestedDataWriteBenchmark.writeFileSource    ss    5  8.798 ± 0.342   s/op
SparkParquetNestedDataWriteBenchmark.writeIceberg       ss    5  8.457 ± 0.282   s/op
```

This change is also critical as Iceberg will collect proper metadata per file (e.g., min/max values). As for now, we only keep track of the number of rows during writes via Iceberg Spark data source.

rdblue commented:
Thanks @aokolnychyi! Looks like it isn't faster, but at least it isn't slower.

aokolnychyi commented:
I will share the benchmarking code later and we can discuss if the test is valid or not. Overall, I saw that the performance on flat data is very close but Iceberg is a bit faster on nested data.

I have more benchmarks for the read path. There are also interesting moments.

mccheah commented on spark/src/main/java/com/netflix/iceberg/spark/data/SparkParquetWriters.java:
```diff
@@ -0,0 +1,447 @@
+    private final int precision;
+    private final int scale;
+    private final int length;
+    private final ThreadLocal<byte[]> bytes;
```
At one point would multiple threads access the same writer? Does this truly need to be a `ThreadLocal`?

mccheah replied:
Wonder if we can achieve something similar by making our own byte buffer pool that's instantiated here. Can make the lifecycle of the reused byte buffers more explicit.

rdblue replied:
These writer classes are used to build a tree that can consume objects that will be written to Parquet files. Those trees are expensive to build because it requires traversing the table schema. That means that we may want to cache and reuse them later, possibly from multiple threads. Because the function they perform is stateless, it would be easy to later assume that these are thread-safe. That's why I went ahead and made them thread-safe.

We could change how this is done, but it is one byte array allocated per thread, so I think this light-weight solution is fine.

mccheah commented on spark/src/main/java/com/netflix/iceberg/spark/data/SparkParquetWriters.java:
```diff
@@ -0,0 +1,447 @@
+    @Override
+    protected Iterator<Map.Entry<K, V>> pairs(MapData map) {
+      return new Iterator<Map.Entry<K, V>>() {
+        private final int size = map.numElements();
```
At least to me it's a little harder to read inner classes with fields if they're anonymous like this. Can we separate this `Iterator` out to its own class?

mccheah commented on spark/src/main/java/com/netflix/iceberg/spark/data/SparkParquetWriters.java:
```diff
@@ -0,0 +1,447 @@
+    @SuppressWarnings("unchecked")
+    protected Iterator<E> elements(ArrayData list) {
+      return new Iterator<E>() {
+        private final int size = list.numElements();
```
Can we separate out this iterator into its own class?

mccheah commented on spark/src/main/java/com/netflix/iceberg/spark/data/SparkParquetWriters.java:
```diff
@@ -0,0 +1,447 @@
+
+    @Override
+    public ParquetValueWriter<?> map(GroupType map,
+                                     ParquetValueWriter<?> keyWriter,
```
Shouldn't indentation for this and many other methods in this class be 4 spaces in from the `public` modifier on the previous line? Also might want `map` to be on its own line. In general the formatting for method declarations throughout seems off.

rdblue replied:
The rule I was using was to indent to the start of arguments, and if any one argument could not fit in the space, to move all arguments back to 4 spaces from the method definition's start. I'm fine with updating this convention.

aokolnychyi replied:
Out of curiosity, how many args do we want to keep per line? I've seen a few args per line in other places.

rdblue replied:
We should standardize this. It would be good to just pick a rule.

aokolnychyi commented:
We tested this change locally and did not notice any issues so far. It would be nice to see this merged once @mccheah's comments are addressed.

prodeezy commented:
@rdblue Thanks for this PR.  I applied this to latest and ran my functional tests. Works great! confirmed that top level stats are kept in Manifests now and filters can prune files :-) .. File pruning is a critical feature for us. Looking forward to seeing this merged. 

P.S. master branch has diverged slightly so the PR patch didn't apply cleanly. Some slight changes were needed. No biggie but wanted to let you know.

prodeezy commented on spark/src/main/java/com/netflix/iceberg/spark/data/SparkParquetWriters.java:
```diff
@@ -0,0 +1,447 @@
+          case INT_8:
+          case INT_16:
+          case INT_32:
+          case INT_64:
```
I realize that INT96 is deprecated as per https://issues.apache.org/jira/browse/PARQUET-323 and don't want to encourage people using it in Iceberg but if people do want to re-write data with INT96 rows into iceberg (as INT64)  would we run into this code? If so should we not handle it by truncating INT96 to INT64?

rdblue replied:
No, by the time this code gets INT96 values, they will be converted to binary, fixed, or a long. That's the engine's job because Iceberg won't be used to read INT96 values from files. So another read implementation produces values, then the engine hands them to Iceberg using a known type. No need to handle it here.

rdblue commented:
@mccheah, @prodeezy, @aokolnychyi, I've rebased and updated this PR for your review comments. Please have another look when you get a chance. Thank you!

prodeezy commented:
Minor nit: Would be nice to update the PR title to also reflect that this change adds top level column metrics to Iceberg.

prodeezy commented on spark/src/main/java/com/netflix/iceberg/spark/data/SparkParquetWriters.java:
```diff
@@ -0,0 +1,469 @@
+
+    @Override
+    public ParquetValueWriter<?> message(MessageType message,
+                                         List<ParquetValueWriter<?>> fieldWriters) {
```
Style: fix indentation to standard

rdblue replied:
There's discussion on this further down. We should decide on a style for long method arguments and add rules to enforce it. I have generally used this style, but I know others have used 4 spaces and one argument per line. Some places where this style can't fit single arguments have used 4 spaces from the method definition and as many arguments per line as possible... so while I would say that this is pretty normal for the codebase, we need to clean it up.

rdblue commented:
> Minor nit: Would be nice to update the PR title to also reflect that this change adds top level column metrics to Iceberg.

@prodeezy: What does this change that affects top-level column metrics? This shouldn't change metrics because they are taken from the footer after the file is written. That doesn't have much to do with how records are deconstructed and written.

#136 addresses the missing nested metrics problem, right?

aokolnychyi commented:
I think @prodeezy meant that the metrics are not actually fetched from the footer as of today. Correct me if I am wrong, but we still use `ParquetWriteAdapter` right now. It doesn't read the footer. 

Anyway, I believe collecting min/max stats is a side-effect of switching to Iceberg writers. The main feature is that we actually switch to Iceberg writers, so it makes sense to keep the PR name as it is.

rdblue commented:
I didn't know that was broken! Thanks for pointing it out. I've opened #137 to fix the bug in the `WriteSupport` path.

rdblue commented:
Thanks for the reviews, everyone! I'm merging this.
